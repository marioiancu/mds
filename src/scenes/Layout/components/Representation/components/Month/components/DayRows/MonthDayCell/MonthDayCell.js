import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid, } from 'semantic-ui-react';
import { Link } from 'react-router-dom'
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { displayEventModal } from '../../../../../../../../.././data/actions';
import HoverableIcon from '../../../../../../../.././components/HoverableIcon/HoverableIcon';


const propTypes = {
  displayYear: PropTypes.number.isRequired,
  displayMonth: PropTypes.number.isRequired,
  // day: PropTypes.oneOf([
  //   PropTypes.number,
  //   null
  // ]), // How to do ProptTypes.onOfType number || null?
  events: PropTypes.array.isRequired,
  displayEventModal: PropTypes.func.isRequired

};

const defaultProps = {};

export class MonthDayCell extends Component {
  constructor(props) {
    super(props);
    this.acceptedDays = [
      1, 2, 3, 4, 5, 6, 7,
      8, 9, 10, 11, 12, 13,
      14, 15, 16, 17, 18, 19,
      20, 21, 22, 23, 24, 25,
      26, 27, 28, 29, 30, 31
    ];
    this.state = {
      isHoveredOver: false,
      eventsPerDayArray: this.eventsPerDayInMonth(this.props.events)
    };
  }
  componentWillReceiveProps(nextProps) {
    this.setState({ eventsPerDayArray: this.eventsPerDayInMonth(nextProps.events) });
  }
  render() {
    const { acceptedDays, cellClasses, cellColor } = this
    const { isHoveredOver, eventsPerDayArray } = this.state;
    const { day, displayYear, displayMonth, displayEventModal, } = this.props;
    return (
      <Grid.Column className={cellClasses(day)} color={cellColor(day)}
        onMouseEnter={this.handleHoverEnter} onMouseLeave={this.handleHoverExit}>
        <Link to={`/day/${displayYear}/${displayMonth + 1}/${day}`} >
          <div>{day}</div>
        </Link>
        {acceptedDays.includes(day) && <HoverableIcon name="add" show={isHoveredOver}
          onClickCb={displayEventModal} cbArgs={{ year: displayYear, month: displayMonth, day }} />}
        {/* Gross, break out a function for this */}
        {eventsPerDayArray[day] > 0
          ? `${eventsPerDayArray[day]} events`
          : ''}
      </Grid.Column>
    );
  }
  //TODO TEST
  eventsPerDayInMonth = (events) => {
    const daysArr = Array.from({ length: 32 }, (_, i) => 0);
    return events.reduce((accum, event) => {
      accum[event.date.day]++;
      return accum;
    }, daysArr);
  }
  // decide if a cell should be greyed out
  cellColor = (day) => {

    if (this.acceptedDays.includes(day))
      return;
    else
      return "grey";
  }

  // gives custom classes to greyed out cells
  // such as background color and a more defined shadow
  cellClasses = (day) => {

    if (this.acceptedDays.includes(day))
      return;
    else
      return "no-day shadow";
  }
  handleHoverEnter = () => {
    this.setState({ isHoveredOver: true });
  }
  handleHoverExit = () => {
    this.setState({ isHoveredOver: false });
  }
}

const mapStateToProps = state => {
  // const {  } = state;
  return {}
}

const mapDispatchToProps = dispatch => bindActionCreators({
  displayEventModal
}, dispatch);

MonthDayCell.propTypes = propTypes;

MonthDayCell.defaultProps = defaultProps;

// export mapDispatchToProps;
// export { MonthDayCell };
// export monthDayCellInstance;
export default connect(mapStateToProps, mapDispatchToProps)(MonthDayCell);
