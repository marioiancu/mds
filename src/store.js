import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import rootReducer from './reducer';
import { loadState, saveState } from './services/localStorage/lsRedux';

// If there is data saved under 'events' in localStorage, use it 
// in initialising the redux store 
const lsSavedEvents = loadState('events');
let defaultState = {};
if (lsSavedEvents) {
  defaultState = {
    layout: {
      representation: {
        data: {
          events: lsSavedEvents
        }
      }
    }
  }
}

const enhancers = compose(
  applyMiddleware(reduxImmutableStateInvariant()),
  window.devToolsExtension ? window.devToolsExtension() : f => f
);

const store = createStore(rootReducer, defaultState, enhancers);

// Save the events to localStorage on any state change
store.subscribe(() => {
  // This stores state on every state change, so it should be debounced
  // if we every add any actions to the app the fire rapidly and 
  // update the store quickly
  saveState(store.getState().layout.representation.data.events);
})

export default store;
