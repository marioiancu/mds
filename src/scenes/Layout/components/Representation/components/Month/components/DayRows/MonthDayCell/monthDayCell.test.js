import React from 'react';
import renderer from 'react-test-renderer';
// import events from '../../../../../../../../.././services/mock/generatedMockEvents3'
import store from '../../../../../../../../.././store';
const events = Object.values(store.getState().layout.representation.data.events.byId);
import MonthDayCell from './MonthDayCell.js';
import { Provider } from 'react-redux';
import { StaticRouter } from 'react-router';
it('MonthDayCell redners properly', () => {
  const tree = renderer.create(
    <Provider store={store} >
      <StaticRouter context={{}} >
        <MonthDayCell displayMonth={2} displayYear={2016} displayEventModa={() => { }}
          events={events}
        />
      </StaticRouter>
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

// const wrapper = shallow(<MonthDayCell />);

// // import { MonthDayCell } from './MonthDayCell.js';
// import { shallow } from 'react-test-renderer';
// import ReactTestRenderer from 'react-test-renderer';

// const wrapper = ReactTestRenderer.create(
//   <Provider store={store}>
//     <MonthDayCell />
//   </Provider>
// );
// const wrapper = ReactTestRenderer.create(<MonthDayCell />);

// import ReactDOM from 'react-dom';
// const wrapper = ReactDOM.render(<MonthDayCell />);
// const wrapper = shallow(<MonthDayCell />);

// const wrapper = jest.mock('./MonthDayCell', () => () => <div />);
// const wrapper = jest.genMockFromModule('./MonthDayCell', () => () => <div />);
// const something = wrapper.default.WrappedComponent.getMockImplementation;

// console.log(wrapper.default.WrappedComponent.getMockImplementation);
// test('Cell color should be default if there is a day in it', () => {
//   expect(cellColor(1)).toBe(undefined);
// });

// test('Cell color should be gray if there is NO day in that cell', () => {
//   expect(cellColor(undefined)).toBe("grey");
// });

// test('Cell color should be gray if there is NO day in that cell', () => {
//   expect(cellColor(-1)).toBe("grey");
// });

// test('Cell color should be gray if there is NO day in that cell', () => {
//   expect(cellColor("asd")).toBe("grey");
// });

// test('Cell should have no classes if it has a day in it', () => {
//   expect(cellClasses(1)).toBe(undefined);
// });

// test('Cell classes should be appropriate if it DOES NOT have a day in it', () => {
//   expect(cellClasses(undefined)).toBe("no-day shadow");
// });

// test('Cell classes should be appropriate if it DOES NOT have a day in it', () => {
//   expect(cellClasses(-1)).toBe("no-day shadow");
// });

// test('Cell classes should be appropriate if it DOES NOT have a day in it', () => {
//   expect(cellClasses("asd")).toBe("no-day shadow");
// });