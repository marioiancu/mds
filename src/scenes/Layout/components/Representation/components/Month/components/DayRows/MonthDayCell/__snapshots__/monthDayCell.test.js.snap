// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthDayCell redners properly 1`] = `
<div
  className="grey column no-day shadow"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <a
    href="/day/2016/3/undefined"
    onClick={[Function]}
  >
    <div />
  </a>
  
</div>
`;
